Add ability for hiding Carousel controls

YuriyKovalcuk opened the pull request:
≈

<!--- Provide a general summary of the PR in the Title above -->

#### What does this PR do?
Add addition properties for hiding Carousel controls.
#### Where should the reviewer start?
#3174 
#### What testing has been done on this PR?

#### How should this be manually tested?
use the storybook
#### Any background context you want to provide?
#3174 
#### What are the relevant issues?

#### Screenshots (if appropriate)

#### Do the grommet docs need to be updated?
yes
#### Should this PR be mentioned in the release notes?
yes 
#### Is this change backwards compatible or is it a breaking change?
non-breaking

ericsoderberghp commented:
Three things:
1) When would the caller want to turn off one of the control surfaces but not the other? I'm just looking for an example to ensure we're thinking about this in the right way.
2) It would be more consistent with the Video component to have a `controls` property. My inclination is to just support `true|false` to start with.
3) This pull request is missing changes to the typescript typings file: `index.d.ts`.

YuriyKovalcuk commented:
@ericsoderberghp thanks for your comments,
1. Actually, for my particular situation, we have a design mockup, where we have to show only arrows for controls, not selectors. Also in my case the number of carousel items is dynamic, so in case we have only one item, we don`t want to show any controls. And actually, I believe it is a quite common situation when we need to use only one of the control surfaces but not the other, UI frameworks as the Bootstrap or UI-materials, have this ability.
2. I agree it is a better solution to be more consistent with our components.

ShimiSun commented:
@YuriyKovalcuk can we update this PR with resolving the conflicts?

YuriyKovalcuk commented:
@ShimiSun I`ve updated it second ago

ShimiSun commented on src/js/components/Carousel/index.d.ts:
```diff
@@ -7,6 +7,7 @@ export interface CarouselProps {
   margin?: "none" | "xxsmall" | "xsmall" | "small" | "medium" | "large" | "xlarge" | {bottom?: "xxsmall" | "xsmall" | "small" | "medium" | "large" | "xlarge" | string,horizontal?: "xxsmall" | "xsmall" | "small" | "medium" | "large" | "xlarge" | string,left?: "xxsmall" | "xsmall" | "small" | "medium" | "large" | "xlarge" | string,right?: "xxsmall" | "xsmall" | "small" | "medium" | "large" | "xlarg ...
   fill?: boolean;
   play?: number;
+  controls?: "true" | "false" | "arrows" | "selectors"
```
thanks for adding the typescript definition, we are trying to maintain alphabetical order, I know the rest of the file isn't perfect but it will be helpful if we can fix it step by step and moving this line to line #6.

YuriyKovalcuk replied:
@ShimiSun yeah sure, sorry I didn`t notice the order, one moment going to correct it

ShimiSun commented:
@YuriyKovalcuk can you create a new story on src/js/components/Carousel/carousel.stories.js, that will demonstrate this behavior and the behavior from PR #3177 , it will help us: 
1. Verify the functionality.
2. Track future regression behavior - if occurred.
3. Share with the community your awesome work :)

YuriyKovalcuk commented:
@ShimiSun sure, have already started :)

YuriyKovalcuk commented:
@ShimiSun could you please take a look my stories are suitable, or I should correct something? thanks

ShimiSun commented:
@YuriyKovalcuk, @dgennetten It will help us grow our thinking of the Carousel component by understanding when, why and what problem is being solved with the design you are suggesting.
Will it be possible to share a mock/wireframe of the component and see what you are trying to achieve?

ShimiSun commented on src/js/components/Carousel/carousel.stories.js:
```diff
@@ -22,4 +38,27 @@ const SimpleCarousel = () => (
 );
 
-storiesOf('Carousel', module).add('Simple Carousel', () => <SimpleCarousel />);
+const CustomCarousel = () => (
```
Overall the direction is great, THANKS!
I will suggest removing the usage of ThemeContext, it's a better practice to use it when wanting to override and apply a custom theme, in this case, we don't really override and can apply the theme on the Grommet component, see my suggested commits

YuriyKovalcuk replied:
@ShimiSun ok, my idea was to show how we can extend the theme, but for sure, it makes more sense to pass the theme into the Grommet Component here

ShimiSun commented on src/js/components/Carousel/carousel.stories.js:
```diff
@@ -22,4 +38,27 @@ const SimpleCarousel = () => (
+const CustomCarousel = () => (
+  <Grommet>
+    <Box align="center" pad="large">
+      <ThemeContext.Extend value={customTheme}>
```
```suggestion
      
```

ShimiSun commented on src/js/components/Carousel/carousel.stories.js:
```diff
@@ -22,4 +38,27 @@ const SimpleCarousel = () => (
+            <Car size="xlarge" />
+          </Box>
+        </Carousel>
+      </ThemeContext.Extend>
```
```suggestion
    
```

ShimiSun commented on src/js/components/Carousel/carousel.stories.js:
```diff
@@ -22,4 +38,27 @@ const SimpleCarousel = () => (
 
-storiesOf('Carousel', module).add('Simple Carousel', () => <SimpleCarousel />);
+const CustomCarousel = () => (
+  <Grommet>
```
```suggestion
  <Grommet theme={customTheme}>
```

YuriyKovalcuk commented:
@ShimiSun I can provide some information about our suggestions:
1. In our case, we have a dynamic number of carousel items, and in case there is only one item, we want to hide all controls, to keep the design simple. I think it can be also pretty useful if you want to build the carousel with play property. Also from my experience, I can say that it is a quite common situation when the user may want to see only one of the control surfaces but not the other. 
2. The duration of animation is also very important for us. In our case, the carousel takes all screen width, so it may take a long time to change the items, so for us, it is very useful to have the ability to configure the duration. About icons color, it is also very useful to have the ability for styling carousel icons(in case let say, we want to have one color global icons and other for carousel icons).
I hope it is helpful for you, please tell me if I can provide something else.

Also, I think @dgennetten  can provide much more information here, He has really huge UI-UX experience.

ShimiSun commented:
@YuriyKovalcuk , great inputs, it will help us consider the enhancements when we discuss the component within the team. Doug and I will follow up as well.

ShimiSun commented:
@YuriyKovalcuk & @dgennetten any news here?

dgennetten commented:
Unfortunately but not surprisingly, CTP have decided to move away from grommet on this one.

ericsoderberghp commented:
I would like to see this come in. I've been using a Carousel in a project and would like to hide the controls so I can just leverage the animation.

dgennetten commented:
yet, y'all pushed back when CTP wanted this to move forward?

YuriyKovalcuk commented:
@ShimiSun, @ericsoderberghp, actually, I fixed and corrected all comments to this PR, added examples to storybook as well. If you have some other comments please let me know, or you can just merge this PR

ShimiSun commented:
@YuriyKovalcuk thanks for collaboration, any chance you can resolve the conflicts?

YuriyKovalcuk commented:
@ShimiSun sure, but can it wait at least one day? Currently, I am on vacation and to be an honest quality of internet here is really bad ;) I have been trying to add this comment during 10 minutes ;)

ShimiSun commented:
@YuriyKovalcuk absolutely, lol, no rush at all. 
First and foremost enjoy your vacation 🏖

YuriyKovalcuk commented:
@ShimiSun conflicts are resolved

ShimiSun commented on src/js/components/Carousel/Carousel.js:
```diff
@@ -81,9 +81,11 @@ class Carousel extends Component {
 
   render() {
-    const { children, fill, focus, theme, ...rest } = this.props;
+    const { children, fill, focus, theme, controls, ...rest } = this.props;
```
```suggestion
    const { children, controls, fill, focus, theme, ...rest } = this.props;
```
we are trying to keep list orders in an alphabetical way :)

ShimiSun commented on src/js/components/Carousel/carousel.stories.js:
```diff
@@ -22,7 +38,28 @@ const SimpleCarousel = ({ initialChild }) => (
+  .add('Simple Carousel', () => <SimpleCarousel />)
+  .add('Without Controls', () => <SimpleCarousel controls={false} play={1000} />)
+  .add('Custom Animation and Styles', () => <CustomCarousel />)
+  .add('Initial child`', () => (
```
```suggestion
  .add('Initial Child', () => (
```
small clean up since we are already here :)

ShimiSun commented on src/js/components/Carousel/carousel.stories.js:
```diff
@@ -22,7 +38,28 @@ const SimpleCarousel = ({ initialChild }) => (
+
+storiesOf('Carousel', module)
+  .add('Simple Carousel', () => <SimpleCarousel />)
+  .add('Without Controls', () => <SimpleCarousel controls={false} play={1000} />)
```
Great story!! thanks buddy! (I'm slowing it down a bit though :) )
```suggestion
  .add('Without Controls', () => <SimpleCarousel controls={false} play={1500} />)
```

YuriyKovalcuk commented:
@ShimiSun, please take a look

ShimiSun commented on src/js/components/__tests__/__snapshots__/README-test.js.snap:
```diff
@@ -2502,6 +2502,16 @@ If specified, the index of
+Whether to show carousel controls and where to place them. Defaults to \`true\`.
+
+\`\`\`
+[object Object]
```
this doesn't look right, it should be fixed with the doc.js definition.

ShimiSun commented on src/js/components/Carousel/doc.js:
```diff
@@ -25,6 +25,9 @@ export const doc = Carousel => {
       will loop through all children indefinitely.`),
     initialChild: PropTypes.number.description(`If specified, the index of
       the first element to be shown. Defaults to 0.`),
+    controls: PropTypes.oneOf([PropTypes.bool, 'arrows', 'selectors'])
```
```suggestion
    controls: PropTypes.oneOf([PropTypes.bool, PropTypes.oneOf(['arrows', 'selectors'])])
```
I'm thinking this will generate a better Readme file and be more proptype accurate.

YuriyKovalcuk replied:
true true, my bad

ShimiSun commented on src/js/components/Carousel/carousel.stories.js:
```diff
@@ -22,7 +56,10 @@ const SimpleCarousel = ({ initialChild }) => (
-  <SimpleCarousel initialChild={1} />
-));
+storiesOf('Carousel', module)
+  .add('Simple Carousel', () => <SimpleCarousel />)
```
```suggestion
  .add('Simple', () => <SimpleCarousel />)
```

ShimiSun commented:
Thank you @YuriyKovalcuk for your contribution!! 🥇
